src/feature/hs/hs_descriptor.c

```diff
@@ -2580,17 +2576,17 @@ hs_desc_decode_descriptor(const char *encoded,
   memcpy(desc->subcredential, subcredential, sizeof(desc->subcredential));
 
   ret = hs_desc_decode_plaintext(encoded, &desc->plaintext_data);
-  if (ret < 0) {
+  if (ret != HS_DESC_DECODE_OK) {
     goto err;
   }
 
   ret = hs_desc_decode_superencrypted(desc, &desc->superencrypted_data);
-  if (ret < 0) {
+  if (ret != HS_DESC_DECODE_OK) {
     goto err;
   }
 
   ret = hs_desc_decode_encrypted(desc, client_auth_sk, &desc->encrypted_data);
-  if (ret < 0) {
+  if (ret != HS_DESC_DECODE_OK) {
```

Comment:
I don't quite understand the code flow here that ends up with us storing the failed-to-decrypt descriptors. If we go to `err` when we fail to decrypt here, and the err frees the descriptor, how do we end up with the descriptor in the cache?

I guess this works in the end because the unittests say so, but I really have trouble seeing this in the code... Is there a way to clarify this in the code with comments or anything?

Reply:
Well I think you are just not seeing the top layer here with the HS cache.

The HS cache (client side), has a top level object that contains the raw blob desc _and_ the decoded descriptor.

In case of a bad client auth, we can NOT have the decoded descriptor which is why this code path returns an error. Remember, everything in `hs_descriptor.c` is about a _decoded_ descriptor object.

So on error of decoding a descriptor, we will keep the undecoded blob in the top level cache object making the "desc == NULL" in that object.